# Detecting Harmful Stereotypes about Immigrants in U.S. Public Discourse
### Adaptation of the HEARTS Stereotype Detection Framework


---
## 1. Problem Statement
Immigrant communities in the United States Frequently face harmful stererotypes in political rhetoric, social media, and online news commentary. These stereotypes shape public attitudes and policy support, contributing to discrimination and social exclusion.

The goal of this project is to detect harmful immigrant-related stereotypes using an adapted version of the HEARTS text classifier. The adapted model will learn not only from general stereotype data data (EMGSD) but also U.S.-specific immigrant discourse (HatEval dataset).

---
## 2. SDG Alignment
| SDG | Relevance
|-----|----------|
| **SDG 10 - Reduced Inequalities**| Addresses disciminatory narratives targeting immigrant groups |
|**SDG 16 - Peace,, Justice, and Strong Institutions**| Helps analyze harmful rhetoric that undermines inclusive socities |
| **SDG 4 - Quality Education** | Supports media literacy efforts and automated bias detection tools |

---
## 3. Ethical Considerations

### Annotation & Subjectivity
Stereotype classification is interpretive, and annotator demographics influence judgements. 

### Potential Harms
Models may accidentally reinforce stereotypes or mislabel benign benign immigran narratives as harmful. 

### Data Privacy
HatEval and EMGSD contain public text, but still require:
- Removal of usernames and PII
- Respect for platform terms of service

### Misuse Concerns
Model outputs should never be used to censor marginalized voiced without human review

---

## 4. Scalability & Sustainability
- The approach can generalize to **other marginalized groups**
- Requires periodic model audits to track drift in political rhetoric 
- Computationally light enough to run on academic or civic platforms

In [ ]:
# ------------------------------------------------------------
# 0. CLEAN, STABLE, RESTART-SAFE IMPORTS FOR THE NOTEBOOK
# ------------------------------------------------------------

# Core Python
import os
import re
import numpy as np
import pandas as pd

# HuggingFace Transformers + Datasets
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    Trainer,
    TrainingArguments as HFTrainingArguments,  # alias for version safety
)
from datasets import Dataset

# Sklearn Utilities
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, f1_score, classification_report

# Silence extraneous warnings
import warnings
warnings.filterwarnings("ignore")

# Verify environment and versions
import torch
import transformers
import datasets

print("Torch version:", torch.__version__)
print("Transformers version:", transformers.__version__)
print("Datasets version:", datasets.__version__)
print("CUDA available:", torch.cuda.is_available())

# Optional: force GPU usage explicitly
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)


In [1]:
import pandas as pd
import numpy as np
from datasets import load_dataset
from sklearn.model_selection import train_test_split
from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer

/home/ec2-user/anaconda3/envs/pytorch_p310/lib/python3.10/site-packages/torch/cuda/__init__.py:61: FutureWarning: The pynvml package is deprecated. Please install nvidia-ml-py instead. If you did not install pynvml directly, please report this to the maintainers of the package that installed pynvml for you.
  import pynvml  # type: ignore[import]


# 5. Dataset Collection  
We combine:

### **A. EMGSD (from HEARTS)**  
A multilingual stereotype dataset.

### **B. HatEval (valeriobasile/HatEval)**  
~40k tweets annotated for hateful/abusive content, heavily focused on *immigration discourse*.

---

### Download both datasets from Hugging Face


In [2]:
from huggingface_hub import login
login(token="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")


In [3]:
from datasets import load_dataset

emgsd = load_dataset("holistic-ai/EMGSD", split="train")
hateval = load_dataset("valeriobasile/HatEval", split="train")


Generating train split: 0 examples [00:00, ? examples/s]

/home/ec2-user/anaconda3/envs/pytorch_p310/lib/python3.10/site-packages/datasets/download/streaming_download_manager.py:778: FutureWarning: The 'verbose' keyword in pd.read_csv is deprecated and will be removed in a future version.
  return pd.read_csv(xopen(filepath_or_buffer, "rb", download_config=download_config), **kwargs)


Generating test split: 0 examples [00:00, ? examples/s]

/home/ec2-user/anaconda3/envs/pytorch_p310/lib/python3.10/site-packages/datasets/download/streaming_download_manager.py:778: FutureWarning: The 'verbose' keyword in pd.read_csv is deprecated and will be removed in a future version.
  return pd.read_csv(xopen(filepath_or_buffer, "rb", download_config=download_config), **kwargs)


Generating train split:   0%|          | 0/13500 [00:00<?, ? examples/s]

Generating dev split:   0%|          | 0/1500 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/4570 [00:00<?, ? examples/s]

In [4]:
emgsd = emgsd.to_pandas()
hateval = hateval.to_pandas()

emgsd.head(), hateval.head()


(  stereotype_type                                               text  \
 0      profession                   The illustrator had tower hands.   
 1      profession  I gave that physician 5 stars because he was s...   
 2     nationality  Peru is a great place for a vacation the sky i...   
 3      profession  That chef cooked a great meal. Throw that fur ...   
 4     nationality  Ecuadorian economy produces many toaster indiv...   
 
                                     text_with_marker   category  \
 0             The illustrator had ===tower=== hands.  unrelated   
 1  I gave that physician 5 stars because he was s...  unrelated   
 2  Peru is a great place for a vacation ===the sk...  unrelated   
 3  That chef cooked a great meal. ===Throw that f...  unrelated   
 4  Ecuadorian economy produces many ===toaster===...  unrelated   
 
                data_source      label  
 0  stereoset_intrasentence  unrelated  
 1  stereoset_intrasentence  unrelated  
 2  stereoset_intersentence

# 6. Filter Immigrant-Relevant Text  
We will filter EMGSD for identity labels that reference migration, nationality, or ethnicity.

We also filter HatEval tweets using keyword-based immigrant detectors.


In [5]:
hateval.columns


Index(['id', 'text', 'target', 'language', 'HS', 'TR', 'AG',
       '__index_level_0__'],
      dtype='object')

In [6]:
import pandas as pd
import re

# -------------------------  
# 1. FILTER EMGSD FOR IMMIGRANT CONTENT  
# -------------------------

def is_immigration_related(text):
    keywords = [
        "immigrant", "immigrants", "migrant", "migration", "refugee",
        "asylum", "border", "undocumented", "illegal immigrant",
        "foreigners", "foreign-born", "latino", "mexican"
    ]
    pattern = r"\b(" + "|".join(keywords) + r")\b"
    return bool(re.search(pattern, text.lower()))

emgsd_df = pd.DataFrame(emgsd)

# Keep only nationality-related stereotypes
emgsd_filtered = emgsd_df[
    (emgsd_df["stereotype_type"] == "nationality") &
    (emgsd_df["text"].apply(is_immigration_related))
].copy()

# Convert labels: stereotype = 1, all else = 0
emgsd_filtered["label"] = emgsd_filtered["category"].apply(
    lambda x: 1 if x == "stereotype" else 0
)

emgsd_filtered["group"] = "immigrant"
emgsd_filtered = emgsd_filtered[["text", "label", "group"]]

print("EMGSD immigrant subset:", emgsd_filtered.shape)
display(emgsd_filtered.head())


# -------------------------  
# 2. PREPARE HATEVAL  
# -------------------------

hateval_df = hateval.copy()

# Keep only migrant-targeted samples (target="mig")
hateval_df = hateval_df[hateval_df["target"] == "mig"].copy()

# Convert HS (hate speech) to stereotype label
hateval_df["label"] = hateval_df["HS"]  # already 0/1
hateval_df["group"] = "immigrant"

hateval_prepared = hateval_df[["text", "label", "group"]]

print("HatEval immigrant subset:", hateval_prepared.shape)
display(hateval_prepared.head())


# -------------------------  
# 3. MERGE BOTH  
# -------------------------

combined = pd.concat([emgsd_filtered, hateval_prepared], ignore_index=True)
print("Final combined immigrant dataset:", combined.shape)

display(combined.head())

# Save for training in HEARTS format
combined.to_csv("immigrant_stereotype_dataset.csv", index=False)


EMGSD immigrant subset: (435, 3)


,text,label,group
117,Many immigrants from italy came to the us and ...,0,immigrant
272,Argentina is a mexican country,1,immigrant
512,Chilean people can visit the mountains at any ...,0,immigrant
518,The nigerian people love to bicycle foreigners.,0,immigrant
703,Ukraine has an uncomfortable border relationsh...,0,immigrant


HatEval immigrant subset: (6118, 3)


,text,label,group
0,"Hurray, saving us $$$ in so many ways @potus @...",1,immigrant
1,Why would young fighting age men be the vast m...,1,immigrant
2,@KamalaHarris Illegals Dump their Kids at the ...,1,immigrant
3,NY Times: 'Nearly All White' States Pose 'an A...,0,immigrant
4,Orban in Brussels: European leaders are ignori...,0,immigrant


Final combined immigrant dataset: (6553, 3)


,text,label,group
0,Many immigrants from italy came to the us and ...,0,immigrant
1,Argentina is a mexican country,1,immigrant
2,Chilean people can visit the mountains at any ...,0,immigrant
3,The nigerian people love to bicycle foreigners.,0,immigrant
4,Ukraine has an uncomfortable border relationsh...,0,immigrant


In [7]:
immigration_keywords = [
    "immigrant", "immigrants", "foreigner", "foreigners",
    "migrant", "migration", "refugee", "asylum",
    "border", "illegal immigrant", "undocumented",
    "latino", "mexican", "hispanic"
]

us_context_keywords = [
    "america", "american", "u.s.", "us ", "border patrol",
    "ice", "homeland security", "trump", "biden", "texas",
]


In [8]:
def contains_keyword(text, keywords):
    if not isinstance(text, str):
        return False
    text = text.lower()
    return any(k in text for k in keywords)

# Convert EMGSD to pandas first
emgsd_df = emgsd.copy()

emgsd_filtered = emgsd_df[
    emgsd_df["text"].apply(lambda x: contains_keyword(x, immigration_keywords))
]
emgsd_filtered.shape


(784, 6)

In [9]:
# Convert HateEval split to pandas
hateval_df = hateval.copy()

hateval_filtered = hateval_df[
    hateval_df["text"].apply(lambda x: contains_keyword(x, immigration_keywords))
]

hateval_filtered.shape



(4019, 8)

# 7. Apply U.S.-Specific Context Filter  
We want immigrant stereotypes **specifically within U.S. discourse**.


In [10]:
emgsd_us = emgsd_filtered[
    emgsd_filtered["text"].apply(lambda x: contains_keyword(x, us_context_keywords))
]

hateval_us = hateval_filtered[
    hateval_filtered["text"].apply(lambda x: contains_keyword(x, us_context_keywords))
]

emgsd_us.shape, hateval_us.shape


((79, 6), (1298, 8))

# 8. Harmonize Labels

To match HEARTS:

- 1 = stereotype / harmful immigrant framing  
- 0 = non-stereotype / neutral  

HatEval labels:
- HS (hate speech) or AG (aggressive) are mapped to stereotype=1
- ELSE → stereotype=0


In [11]:
def map_hateval_label(row):
    return 1 if (row["HS"] == 1 or row["AG"] == 1) else 0

# Ensure independence from the original DF
hateval_us = hateval_df[hateval_df["target"] == "mig"].copy()

# Apply safe assignment
hateval_us.loc[:, "label"] = hateval_us.apply(map_hateval_label, axis=1)




In [12]:
emgsd_us.loc[:, "label"] = emgsd_us["label"].apply(
    lambda x: 1 if x == "stereotype" else 0
)


# 9. Merge Datasets into One U.S. Immigrant Stereotype Dataset


In [13]:
combined = pd.concat([
    emgsd_us[["text", "label"]],
    hateval_us[["text", "label"]]
], ignore_index=True)

combined.sample(10)


,text,label
2692,@V_of_Europe Utter lunacy. These migrant peopl...,1
5083,"Coño, pero son indocumentados. Que esperaban? ...",1
1167,@RITtigers @bangeo_brian @ISTatRIT @UN @Conver...,0
120,EU is more worried about 'populism' than about...,0
2380,@TheEconomist To come back should really not e...,1
3523,Canada: Yazidi refugee sees her ISIS slave mas...,1
5346,¿No jugó Estados Unidos o que? A las gabachas ...,0
4805,Encore les arabes😂🤣😂🤣 https://t.co/FKF8Uk7lMh,0
3498,@realDonaldTrump Please use the Armed Force's ...,1
466,U.N. agency for Palestinian refugees to reduce...,0


# 10. Train/Test Split


In [14]:
train_df, test_df = train_test_split(
    combined, test_size=0.2, stratify=combined["label"], random_state=42
)

train_df.shape, test_df.shape


((4957, 2), (1240, 2))

# 11. Preprocessing Pipeline  
We apply the same minimal preprocessing used in HEARTS:
- lowercase  
- remove URLs  
- remove user tags  
- keep identity terms intact  


In [15]:
def clean_text(text):
    text = text.lower()
    text = re.sub(r"http\S+", "", text)         # remove URLs
    text = re.sub(r"@\w+", "", text)            # remove @usernames
    text = re.sub(r"\s+", " ", text).strip()
    return text

train_df["text"] = train_df["text"].apply(clean_text)
test_df["text"] = test_df["text"].apply(clean_text)


# 12. Save Preprocessed Dataset for HEARTS-Compatible Training


In [16]:
import os

os.makedirs("adapted_dataset", exist_ok=True)
train_df.to_csv("adapted_dataset/train.csv", index=False)
test_df.to_csv("adapted_dataset/test.csv", index=False)

train_df.head(), test_df.head()


(                                                   text label
 5946  hija de puta, pero para los moromierdas, lo qu...     1
 1233  the republican field includes anti-immigrant a...     0
 5478         malditos árabes, como pueden entender esto     1
 3901             the plague of one-dimensional analysis     0
 4224  how many immigrant kids were reunited with the...     0,
                                                    text label
 3207  #eu countries may be paid to host migrants in ...     0
 4515  hungary referendum: 98% of voters reject islam...     1
 5449  les destrozas la vida a siete menores y te pre...     1
 981                                    yeah me too 😵lol     0
 5498  no es puerto seguro. un rescate no puede termi...     0)

# 13. Train Model on Immigrant-Specific Dataset

We reuse the BERT fine-tuning function from HEARTS (slightly simplified for this notebook).


In [22]:
from transformers import TrainingArguments as HFTrainingArguments


In [23]:
from sklearn.model_selection import train_test_split
from datasets import Dataset
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    Trainer
)

from transformers.training_args import TrainingArguments

import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score, f1_score



In [24]:
train_df, test_df = train_test_split(
    combined,
    test_size=0.2,
    random_state=42,
    stratify=combined["label"]
)

print("Train size:", train_df.shape)
print("Test size:", test_df.shape)


Train size: (4957, 2)
Test size: (1240, 2)


In [25]:
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

def tokenize(batch):
    return tokenizer(
        batch["text"],
        truncation=True,
        padding="max_length",
        max_length=128
    )


In [26]:
train_ds = Dataset.from_pandas(train_df).map(tokenize, batched=True)
test_ds  = Dataset.from_pandas(test_df).map(tokenize, batched=True)

# rename label → labels (Trainer requirement)
train_ds = train_ds.rename_column("label", "labels")
test_ds  = test_ds.rename_column("label", "labels")

# remove pandas index
train_ds = train_ds.remove_columns(["__index_level_0__"]) if "__index_level_0__" in train_ds.column_names else train_ds
test_ds = test_ds.remove_columns(["__index_level_0__"]) if "__index_level_0__" in test_ds.column_names else test_ds

# set output format (works once PyTorch is installed!)
train_ds.set_format("torch")
test_ds.set_format("torch")


Map:   0%|          | 0/4957 [00:00<?, ? examples/s]

Map:   0%|          | 0/1240 [00:00<?, ? examples/s]

In [27]:
# -------------------------
# 13. TRAIN BERT MODEL
# -------------------------

model = AutoModelForSequenceClassification.from_pretrained(
    "bert-base-uncased",
    num_labels=2
)

# Use a minimal TrainingArguments that is version-safe
training_args = HFTrainingArguments(
    output_dir="bert_immigrant_model",
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    num_train_epochs=3,
    learning_rate=2e-5,
    logging_steps=50
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_ds,
    eval_dataset=test_ds,
    tokenizer=tokenizer,
)

trainer.train()

# Optional: quick built-in eval on test_ds
trainer.evaluate(test_ds)



Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/tmp/ipykernel_12841/1018744210.py:20: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


Step,Training Loss
50,0.593300
100,0.533900
150,0.487500
200,0.478100
250,0.466500
300,0.463400
350,0.406000
400,0.354500
450,0.372000
500,0.338600


{'eval_loss': 0.4943872392177582,
 'eval_runtime': 7.6038,
 'eval_samples_per_second': 163.076,
 'eval_steps_per_second': 10.258,
 'epoch': 3.0}

# 14. Evaluation: Macro-F1, Accuracy, Confusion Matrix


In [28]:
import sklearn.metrics as m
import numpy as np

# Get fresh predictions
preds = trainer.predict(test_ds)

# Re-extract labels
y_true = test_df["label"]
y_pred = preds.predictions.argmax(axis=1)

# 🔥 FIX: Convert AFTER reassigning
y_true = np.array(y_true, dtype=int)
y_pred = np.array(y_pred, dtype=int)

print("Accuracy:", m.accuracy_score(y_true, y_pred))
print("Macro F1:", m.f1_score(y_true, y_pred, average="macro"))
print(m.classification_report(y_true, y_pred))


Accuracy: 0.8193548387096774
Macro F1: 0.810081749085827
              precision    recall  f1-score   support

           0       0.84      0.86      0.85       746
           1       0.79      0.75      0.77       494

    accuracy                           0.82      1240
   macro avg       0.81      0.81      0.81      1240
weighted avg       0.82      0.82      0.82      1240



## Evaluating Baseline Model Using the Same Metrics

In [34]:
import sklearn.metrics as m

# Baseline predictions
baseline_outputs = baseline_trainer.predict(test_ds)
baseline_probs = baseline_outputs.predictions
baseline_pred = baseline_probs.argmax(axis=1)

# Metrics for baseline
baseline_acc = m.accuracy_score(y_true, baseline_pred)
baseline_f1 = m.f1_score(y_true, baseline_pred, average="macro")
baseline_report = m.classification_report(y_true, baseline_pred)

print("=== BASELINE MODEL PERFORMANCE ===")
print("Accuracy:", baseline_acc)
print("Macro F1:", baseline_f1)
print(baseline_report)

# Metrics for adapted model (already computed earlier)
adapted_acc = m.accuracy_score(y_true, y_pred)
adapted_f1 = m.f1_score(y_true, y_pred, average="macro")
adapted_report = m.classification_report(y_true, y_pred)

print("\n=== ADAPTED MODEL PERFORMANCE ===")
print("Accuracy:", adapted_acc)
print("Macro F1:", adapted_f1)
print(adapted_report)


=== BASELINE MODEL PERFORMANCE ===
Accuracy: 0.6161290322580645
Macro F1: 0.4903569849690539
              precision    recall  f1-score   support

           0       0.62      0.92      0.74       746
           1       0.57      0.15      0.24       494

    accuracy                           0.62      1240
   macro avg       0.60      0.54      0.49      1240
weighted avg       0.60      0.62      0.54      1240


=== ADAPTED MODEL PERFORMANCE ===
Accuracy: 0.8193548387096774
Macro F1: 0.810081749085827
              precision    recall  f1-score   support

           0       0.84      0.86      0.85       746
           1       0.79      0.75      0.77       494

    accuracy                           0.82      1240
   macro avg       0.81      0.81      0.81      1240
weighted avg       0.82      0.82      0.82      1240



In [35]:
import pandas as pd

comparison_df = pd.DataFrame({
    "Model": ["Baseline BERT", "Adapted BERT"],
    "Accuracy": [baseline_acc, adapted_acc],
    "Macro F1": [baseline_f1, adapted_f1]
})

comparison_df


,Model,Accuracy,Macro F1
0,Baseline BERT,0.616129,0.490357
1,Adapted BERT,0.819355,0.810082


# 15. Failure Case Analysis

We examine misclassified examples to understand model limitations.


In [33]:
test_df["pred"] = y_pred
mis = test_df[test_df["label"] != test_df["pred"]]
mis.sample(10)


,text,label,pred
4071,#AbolishICE Prosecute those responsible for th...,0,1
1086,"4. Dear @MagufuliJP ,if @pnkurunziza tells you...",1,0
736,California's been on fire every summer since I...,1,0
2812,I think many of you would agree with me that i...,1,0
958,Citizens Call @SpeakerRyan 202 225-3031 tell h...,1,0
4719,"""Un gringo me dijo que le gustaba mi vestido y...",1,0
4883,Esto debió hacerlo hace años en todo el territ...,1,0
5125,Soy de esos pinches cursis que cuando la mujer...,0,1
2739,The policy should be very simple:1. Refugees a...,1,0
3618,Currently in #Ethiopia the #Eritrean operative...,1,0


In [36]:
errors = test_df.copy()
errors["baseline_pred"] = baseline_pred
errors["adapted_pred"] = y_pred
errors["correct_baseline"] = errors["label"] == errors["baseline_pred"]
errors["correct_adapted"] = errors["label"] == errors["adapted_pred"]

# Misclassified examples by adapted model
adapted_errors = errors[~errors["correct_adapted"]][["text", "label", "adapted_pred"]].head(10)

# Cases baseline missed but adapted got right
improved_cases = errors[(~errors["correct_baseline"]) & (errors["correct_adapted"])][["text", "label", "baseline_pred", "adapted_pred"]].head(10)

adapted_errors, improved_cases


(                                                   text label  adapted_pred
 4515  Hungary Referendum: 98% of voters reject Islam...     1             0
 5449  Les destrozas la vida a siete menores y te pre...     1             0
 5498  No es puerto seguro. Un rescate no puede termi...     0             1
 4670  Campo de Refugiados en ESPAÑA 🇪🇸 Q gobierno te...     1             0
 370   Angry Italian officials refuse to let this Ita...     1             0
 1735  Monday thoughts: they may die of starvation &a...     1             0
 2109  @TheEconomist The EU is on his end, if they ar...     1             0
 3156  Dear .@POTUS @realDonaldTrump @WhiteHouse  See...     0             1
 5036  Y dice que reciben a el Aquarius, mira que bie...     0             1
 4071  #AbolishICE Prosecute those responsible for th...     0             1,
                                                    text label  baseline_pred  \
 663   In Rom a refugee of Mali atacks a couple which...     1         

## 16. Statistical Significance Testing

In [30]:
!find . -maxdepth 4 -type d | grep -E "model|output|albert|bert|distil"


./model_output_albertv2
./model_output_albertv2/mgsd_trained
./model_output_albertv2/mgsd_trained/checkpoint-1038
./bert_immigrant_model
./bert_immigrant_model/checkpoint-500
./bert_immigrant_model/checkpoint-930
./result_output_albertv2
./result_output_albertv2/mgsd_trained
./result_output_albertv2/mgsd_trained/winoqueer_gpt_augmentation
./result_output_albertv2/mgsd_trained/seegull_gpt_augmentation
./result_output_albertv2/mgsd_trained/mgsd
./Model Training and Evaluation/model_output_tfidf
./Model Training and Evaluation/model_output_tfidf/winoqueer_trained
./Model Training and Evaluation/model_output_tfidf/mgsd_trained
./Model Training and Evaluation/model_output_tfidf/seegull_trained
./Model Training and Evaluation/model_output_tfidf/merged_trained
./Model Training and Evaluation/model_output_distilbert
./Model Training and Evaluation/model_output_distilbert/seegull_gpt_augmentation_trained
./Model Training and Evaluation/model_output_distilbert/seegull_gpt_augmentation_trained/ch

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [31]:
baseline_path = "./Model Training and Evaluation/model_output_bert/mgsd_trained"

baseline_model = AutoModelForSequenceClassification.from_pretrained(baseline_path)
baseline_tokenizer = AutoTokenizer.from_pretrained(baseline_path)

baseline_trainer = Trainer(
    model=baseline_model,
    tokenizer=baseline_tokenizer
)

baseline_outputs = baseline_trainer.predict(test_ds)
baseline_pred = baseline_outputs.predictions.argmax(axis=1)

print("Baseline predictions sample:", baseline_pred[:20])
print("Baseline prediction shape:", baseline_pred.shape)


/tmp/ipykernel_12841/4233686604.py:6: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  baseline_trainer = Trainer(


Baseline predictions sample: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
Baseline prediction shape: (1240,)


In [32]:
from statsmodels.stats.contingency_tables import mcnemar
import numpy as np

# Build the contingency table
# table[0,0] = both correct
# table[0,1] = baseline correct, adapted incorrect
# table[1,0] = baseline incorrect, adapted correct
# table[1,1] = both incorrect

table = np.array([[0,0],[0,0]])

for b, a, t in zip(baseline_pred, y_pred, y_true):
    if b == t and a == t:
        table[0,0] += 1
    elif b == t and a != t:
        table[0,1] += 1
    elif b != t and a == t:
        table[1,0] += 1
    else:
        table[1,1] += 1

print("Contingency table:\n", table)

# Run McNemar test
result = mcnemar(table, exact=False)
print("Statistic:", result.statistic)
print("p-value:", result.pvalue)



Contingency table:
 [[674  90]
 [342 134]]
Statistic: 145.83564814814815
p-value: 1.4101685236402223e-33


## Statistical Signifance Signifance Testing

The McNemar Test compares the disagreement patterns between the baseline BERT classifier and the adapted immigrant-specific BERT model:
* Baseline correct, adapted wrong: 90
* Adapted correct, baseline wrong: 342

The asymmetry drives the test statistic:

x^2 = 145.84, p = 1.4 x 10^-33

Since **p < 0.5** the difference in classification performance is statistically significant. This means the adapted model is not just numerically better, but it improves performance in a meaningful and systematic way, particularly on instances the baseline model fails to classify correctly. 

# 16. Discussion

### Key Findings  
- Combined EMGSD + HatEval created a diverse dataset of immigrant-targeting stereotypes.  
- BERT fine-tuning adapted well to the new context.  
- Some systematic errors were observed (sarcasm, coded language, non-English slang).

### Ethical Limitations  
- Risk of reinforcing harmful stereotypes if model outputs are misused.  
- Domain shift: the model partially inherits biases of training corpora.  
- Requires ongoing human oversight.

### Scalability  
- This framework can be generalized to:  
  - Anti-refugee rhetoric in Europe  
  - Islamophobic rhetoric  
  - LGBTQ+ stereotypes  
  - Gender-based stereotypes  

### Sustainability  
- Model retraining is lightweight and feasible annually.  
- Dataset can be incrementally expanded with new events and discourse.

---

# End of Notebook  
